src/adapters/firebase.js

```diff
@@ -0,0 +1,109 @@
+         * @param options.fileName          file name to be set 
+         * @param options.dispatch          function to dispatch events during this process 
+         * @param options.dispatchType      type defining progress of uploading a file
+         * @param options.afterUploaded     async function to execute 
+         */
+        upload (file, options) {
+            assertParam(options, "fileName")  // unless there's no file.filename
+            
+            const {
+                dispatch,
+                dispatchType,
+                afterUploaded,
+                fileName
+            } = options
+
+            const relativePath = path.normalize(`${options.path}/`)
+
+            const ref = firebaseClient.storage().ref(`${relativePath}${fileName}`)
+            const uploadTask = ref.put(file)
+
+            if (dispatch !== undefined && dispatchType !== undefined)
+                uploadTask.on('state_changed', (snapshot) => {
+                    let progress = parseInt((snapshot.bytesTransferred / snapshot.totalBytes) * 100)
+                    dispatch({
+                        type : dispatchType,
+                        data: {
+                            progress
+                        }
+                    })
+                })
+            
+            if (afterUploaded !== undefined) {
+                try {
+                    return uploadTask.then(async (snapshot) => {
+                        await afterUploaded(snapshot)
+                        .then(
+                            console.log('Upload process is done.')
+                        )
+                    })
+                } catch (e) {
+                    console.error(e)
+                }
+            }
+        },
+
+
+        /** 
+         * download file on Firebase storage.
+         * @param storagePath           Firebase storage path.
+         * @param options.destPath      local path to be downloaded.
+         */
+        async download (storagePath, options) {
+
+            const destPath = options.destPath
+
+            const url = await firebaseClient.storage().ref(storagePath).getDownloadURL()
+            const file = fs.createWriteStream(destPath)
+
+            return new Promise ((resolve, reject) => {
+                const request =  https.get(url, (response) => {
+                    if (response.statusCode !== 200) {
+                        reject(new Error('Failed to request'))
+                    }
+                    response.pipe(file)
+                    file.on('finish', () => {
+                        file.close()
+                        resolve('')
+                    })
+                    file.on('error', (err) => {
+                        fs.unlink(destPath, () => {})
+                    })
+                })
+                request.on('error', (err) => {
+                    fs.unlink(destPath, () => {})
+                    reject(err)
+                })
+            })
+
```

Comment:
redundant line?

Reply:
Yeap! I remove all redundant lines I found in other files.